lib/context_adapter_communication.js

```diff
@@ -539,40 +593,94 @@ function notifyOperationCompleted(request, data, callback) {
-          value: getResult(null, data)
+          value: getResult(null, thirdPartyUpdateRequest)
         }
       ]
     }
   ];
 
-  if (request && request.contextAdapter.operationDescriptor.interactionType !==
-    caConfig.BUTTON_ENTITY.INTERACTION_TYPES.ASYNCHRONOUS) {
-    contextElements[0].attributes.push({
-      name: caConfig.BUTTON_ENTITY.IOTA_OPERATION_RESULT_ATTR_NAME,
-      type: caConfig.ATTRIBUTE_TYPE.STRING,
-      value: getResult(null, data)
-    });
+  /**
+   * Removes the op_status and includes the aux_op_result attributes to the array of attributes to send to the
+   *  Context Broker in the updateContext request for the synchronous button case
+   */
+  function setSynchronousButtonAttributes() {
+    contextElements[0].attributes.push(
+      {
+        name: caConfig.BUTTON_ENTITY.IOTA_OPERATION_RESULT_ATTR_NAME,
+        type: caConfig.ATTRIBUTE_TYPE.STRING,
+        value: getResult(null, thirdPartyUpdateRequest)
+      }
+    );
+    caHelper.removeAttribute(contextElements[0].attributes, caConfig.BUTTON_ENTITY.OPERATION_STATUS_ATTR_NAME);
   }
 
-  var payload = {
-    contextElements: contextElements,
-    updateAction: 'UPDATE'
-  };
+  /**
+   * Sends the prepared updateContext request to the Context Broker
+   */
+  function doSendUpdateContext() {
+    var payload = {
+      contextElements: contextElements,
+      updateAction: 'UPDATE'
+    };
 
-  caLogger.debug(
-    request ? request.contextAdapter.context : data.contextAdapter.context,
-    'Notifying the Context Broker that the request has been completed with headers ' +
-    'Fiware-Service: ' + (request ? request.headers['fiware-service'] : data.headers['fiware-service']) + ' and ' +
-    'Fiware-ServicePath: ' + (request ? request.headers['fiware-servicepath'] : data.headers['fiware-servicepath']) +
-    ' and payload: ' + JSON.stringify(payload)
-  );
+    caLogger.debug(
+      request ? request.contextAdapter.context : thirdPartyUpdateRequest.contextAdapter.context,
+      'Notifying the Context Broker that the request has been completed with headers ' +
+      'Fiware-Service: ' + (request ? request.headers['fiware-service'] :
+        thirdPartyUpdateRequest.headers['fiware-service']) + ' and ' +
+      'Fiware-ServicePath: ' + (request ? request.headers['fiware-servicepath'] :
+        thirdPartyUpdateRequest.headers['fiware-servicepath']) +
+      ' and payload: ' + JSON.stringify(payload)
+    );
 
-  sendUpdateContext(
-    request ? request.headers['fiware-service'] : data.headers['fiware-service'],
-    request ? request.headers['fiware-servicepath'] : data.headers['fiware-servicepath'],
-    payload,
-    callback
-  );
+    sendUpdateContext(
+      request ? request.headers['fiware-service'] : thirdPartyUpdateRequest.headers['fiware-service'],
+      request ? request.headers['fiware-servicepath'] : thirdPartyUpdateRequest.headers['fiware-servicepath'],
+      payload,
+      callback
+    );
+  }
+
+  if (request) {
```

Comment:
This is a bit strange. I mean the use of a parameter to decide wether we are answering a syncrhonous or asynchronous request, inside this function. I think each branch of this if should be a sepparate function, invoking different `notifyOperationCompletedSync()` and `notifyOperationCompletedAsync()` functions depending on the type of call (and I assume you know what type of call you are doing, when you decide if you send a `request` object or a `null` value, in lines L182 and L276 of `context_adapter_server.js`). The common behaviour can be extracted to a decorator function that could be called in the asignation of the function to the exports attribute, i.e.: something like this
```
  getServiceDescriptor: getServiceDescriptor,
  sendThirdPartyRequest: sendThirdPartyRequest,
  notifyOperationCompletedSync: notifyOperationCompleted(doSyncOp),
  notifyOperationCompletedAsync: notifyOperationCompleted(doAsyncOp),
  notifyOperationClosed: notifyOperationClosed,
```